### PR TITLE
Metrics Collection System

### PR DESCRIPTION
**Pull request by michael-loki**
3 commits (developed in sync, just grouped some what logically).

1. Adds some useful utilities 
2. Adds metric collection mechanisms 
3. Sample integration with lokinet:
  - Instantiates a metric manager in main
  - Adds a publisher scheduler in main
  - Creates a streaming publisher to `stdout` 
  - Adds some initial collection work to `router/router.cpp`

Work to do (but this should still be mergeable in this state):
- empty metric GC (once a metric is posted once, we always keep track of it)
- json dumper / rotator so we don't fill everyone's disks up
- a remote metrics collector for debugging for lokinet devs (pushing to metric tank or something)
- adding them in more places in the code (the current integration is just sample)

### CODE REVIEW

**Comment by michael-loki** [comment]
edit: this fails debian builds

File: llarp/util/metrics_types.hpp [hunk 1]
```diff
@@ -0,0 +1,532 @@
+#ifndef LLARP_METRICS_TYPES_HPP
+#define LLARP_METRICS_TYPES_HPP
+
+#include <util/string_view.hpp>
+#include <util/threading.hpp>
+#include <util/types.hpp>
+
+#include <absl/types/span.h>
+#include <iosfwd>
+
```
Comment (despair86) [inline llarp/util/metrics_types.hpp]: include what you use:
`absl::optional` (falls through to `std::optional` in modern c++)
`cstring`
`set`
`memory`
to fix non-macOS

Comment (michael-loki) [reply llarp/util/metrics_types.hpp]: yeah, just fixing up debian. looks like some of that is running into a fun gcc constexpr bug though.

include what you use is a good move, I'm bad at remembering.

Comment (michael-loki) [reply llarp/util/metrics_types.hpp]: Fixed.

**Comment by michael-loki** [comment]
Debian is now happy.

**Comment by majestrate** [comment]
merge?

**Comment by michael-loki** [comment]
Go for it, this is into staging not master which makes sense to me.

**Pull request merged**